File: python/graph/graph.py
import sys


class Graph:
    def __init__(self, n, directed=False):
        self.n = n
        self.directed = directed
        self.vertices = [[] for i in range(n)]

    def get_vertices_num(self):
        return self.n

    def get_vertices(self, u):
        return self.vertices[u]

    def is_directed(self):
        return self.directed

    def is_edge(self, u, v):
        if u > self.n or v > self.n:
            return False

        return any(v_ == v for (v_, w_) in self.get_vertices(u))

    def add_edge(self, u, v, w=1):
        if u > self.n or v > self.n:
            raise Exception("Graph.addEdge given vertex larger than size")

        edge = (v, w)
        self.vertices[u].append(edge)

        if not self.directed:
            edge_r = (u, w)
            self.vertices[v].append(edge_r)

    def remove_edge(self, u, v):
        self.vertices[u] = filter(lambda x: x[0] != v, self.get_vertices(u))

    def get_edge_weight(self, u, v):
        w_list = [w_ for (v_, w_) in self.get_vertices(u) if v_ == v]
        return w_list[0] if len(w_list) != 0 else -sys.sys.maxsize

File: scala/src/main/scala/graph/Graph.scala
package graph;

class Graph(val n: Int, val directed: Boolean) {
  var vertices: Array[Seq[(Int, Int)]] = Array.fill(n)(Seq());

  def getVertices(u: Int): Seq[(Int, Int)] = vertices(u);

  def getEdgeWeight(u: Int, v: Int): Int = {
    vertices(u).find(p => p._1 == v) match {
      case Some(pair) => pair._2
      case None       => Int.MinValue
    }
  }

  def isEdge(u: Int, v: Int): Boolean = {
    if (u > n || v > n) false else getVertices(u).exists(p => p._1 == v);
  }

  def addEdge(u: Int, v: Int, w: Int = 1) = {
    if (u > n || v > n)
      throw new Exception("Graph.addEdge given vertex larger than graph size");

    val edge: (Int, Int) = (v, w);
    vertices.update(u, getVertices(u) :+ edge);

    if (!directed) {
      val edgeR: (Int, Int) = (u, w);
      vertices.update(v, getVertices(v) :+ edgeR);
    }
  }

  def removeEdge(u: Int, v: Int) = {
    vertices.update(u, getVertices(u).filter(p => p._1 != v))
  }
}
